forms/FormField.php

```diff
@@ -310,199 +437,251 @@ public function addExtraClass($class) {
 			'id' => $this->ID(),
 			'disabled' => $this->isDisabled(),
 			'readonly' => $this->isReadonly()
 		);
-		
-		if ($this->Required()) {
-			$attrs['required'] = 'required';
-			$attrs['aria-required'] = 'true';
+
+		if($this->Required()) {
+			$attributes['required'] = 'required';
+			$attributes['aria-required'] = 'true';
 		}
 
-		return array_merge($attrs, $this->attributes);
+		return array_merge($attributes, $this->attributes);
 	}
 
 	/**
-	 * @param Array Custom attributes to process. Falls back to {@link getAttributes()}.
-	 * If at least one argument is passed as a string, all arguments act as excludes by name.
-	 * @return string HTML attributes, ready for insertion into an HTML tag
+	 * Custom attributes to process. Falls back to {@link getAttributes()}.
+	 *
+	 * If at least one argument is passed as a string, all arguments act as excludes, by name.
+	 *
+	 * @param array $attributes
+	 *
+	 * @return string
 	 */
-	public function getAttributesHTML($attrs = null) {
-		$exclude = (is_string($attrs)) ? func_get_args() : null;
+	public function getAttributesHTML($attributes = null) {
+		$exclude = null;
+
+		if(is_string($attributes)) {
+			$exclude = func_get_args();
+		}
+
+		if(!$attributes || is_string($attributes)) {
+			$attributes = $this->getAttributes();
+		}
 
-		if(!$attrs || is_string($attrs)) $attrs = $this->getAttributes();
+		$attributes = (array) $attributes;
 
-		// Remove empty
-		$attrs = array_filter((array)$attrs, function($v) {
+		$attributes = array_filter($attributes, function ($v) {
 			return ($v || $v === 0 || $v === '0');
-		}); 
+		});
 
-		// Remove excluded
-		if($exclude) $attrs = array_diff_key($attrs, array_flip($exclude));
+		if($exclude) {
+			$attributes = array_diff_key(
+				$attributes,
+				array_flip($exclude)
+			);
+		}
 
-		// Create markkup
 		$parts = array();
-		foreach($attrs as $name => $value) {
-			$parts[] = ($value === true) ? "{$name}=\"{$name}\"" : "{$name}=\"" . Convert::raw2att($value) . "\"";
+
+		foreach($attributes as $name => $value) {
+			if($value === true) {
+				$parts[] = sprintf('%s="%s"', $name, $name);
+			} else {
+				$parts[] = sprintf('%s="%s"', $name, Convert::raw2att($value));
+			}
 		}
 
 		return implode(' ', $parts);
 	}
 
 	/**
-	 * Returns a version of a title suitable for insertion into an HTML attribute
+	 * Returns a version of a title suitable for insertion into an HTML attribute.
+	 *
+	 * @return string
 	 */
 	public function attrTitle() {
 		return Convert::raw2att($this->title);
 	}
+
 	/**
-	 * Returns a version of a title suitable for insertion into an HTML attribute
+	 * Returns a version of a title suitable for insertion into an HTML attribute.
+	 *
+	 * @return string
 	 */
 	public function attrValue() {
 		return Convert::raw2att($this->value);
 	}
-	
+
 	/**
 	 * Set the field value.
-	 * 
+	 *
 	 * @param mixed $value
-	 * @param mixed $data Optional data source passed in by {@see Form::loadDataFrom}
```

Comment:
I hate to say this but this parameter is super important; We need to add it in formally in 4.0, but we can't remove the informal existence of this variable at this stage.

See my note on https://github.com/silverstripe/silverstripe-framework/pull/4258 where I very nearly did this but decided to wait until a separate PR. :)